File: 438_copy-books-ii/copy-books-ii.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   frankobe
@Problem:  http://www.lintcode.com/problem/copy-books-ii
@Language: Python
@Datetime: 16-11-21 07:13
'''

class Solution:
    # @param n: an integer
    # @param times: a list of integers
    # @return: an integer
    def copyBooksII(self, n, times):
        # write your code here
        if n <= 0 or not times:
            return 0
            
        # dp func: dp[i][j] = min(dp[i][j-1], max(dp[k][j-1], (i-k)*times[j-1])
        # find the optimized k
        # Solution 1: DP TLE            
        # m = len(times)
        # dp = [[0 for i in xrange(m+1)] for j in xrange(n+1)]
        
        # import sys
        # for i in xrange(1,n+1):
        #     dp[i][0] = sys.maxint
        #     dp[i][1] = i*times[0]
            
        # for j in xrange(1, m+1):
        #     dp[1][j] = min(times[:j])
        

        # for j in xrange(2, m+1):
        #     k = 1 
        #     for i in xrange(2, n+1):
        #         dp[i][j] = dp[i][j-1]
        #         while k < i+1:
        #             if dp[i-k][j-1] > k*times[j-1]:
        #                 dp[i][j] = min(dp[i][j], dp[i-k][j-1])
        #                 k += 1
        #             else:
        #                 dp[i][j] = min(dp[i][j], k*times[j-1])
        #                 if k > 1:
        #                     k -= 1
        #                 break
        # return dp[n][m]
        
        # BST
        import heapq
        h = []
        for a in times:
            heapq.heappush(h, (a, a))
        
        for i in xrange(n):
            total, cost = heapq.heappop(h)
            heapq.heappush(h, (total+cost, cost))
            
        result = -1
        for (total, cost) in h:
            if total > cost:
                result = max(result, total-cost)
            
        return result
        
        
        

File: 438_copy-books-ii/copy-books-ii.java
/*
@Copyright:LintCode
@Author:   frankobe
@Problem:  http://www.lintcode.com/problem/copy-books-ii
@Language: Java
@Datetime: 16-11-14 23:48
*/

public class Solution {
    /**
     * @param n: an integer
     * @param times: an array of integers
     * @return: an integer
     */
    public int copyBooksII(int n, int[] times) {
        // write your code here
        int k = times.length;
        int[][] f = new int[2][n+1];
        for (int j = 0 ; j <= n; ++j) {
            f[0][j] = j * times[0];
        }
        for (int i = 1; i < k; ++i) {
            for (int j = 1; j <= n; ++j) {
                int a = i%2;
                int b = 1-a;
                
                f[a][j] = Integer.MAX_VALUE;
                for (int l = 0; l <= j; ++l) {
                    if (f[b][j-l] > times[i] * l) {
                        f[a][j] = Math.min(f[a][j], f[b][j-l]);
                    } else {
                        f[a][j] = Math.min(f[a][j], times[i] * l);
                        break;
                    }
                }
                
            }
        }
        return f[(k-1)%2][n];
    }
}